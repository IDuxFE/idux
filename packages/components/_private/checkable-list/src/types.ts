/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
