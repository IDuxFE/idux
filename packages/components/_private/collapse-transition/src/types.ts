/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
