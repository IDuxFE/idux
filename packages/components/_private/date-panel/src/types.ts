/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export const datePanelProps = {
  cellTooltip: Function as PropType<(cell: { value: Date; disabled: boolean }) => string | void>,
  disabledDate: Function as PropType<(date: Date) => boolean>,
  type: {
    type: String as PropType<DatePanelType>,
    default: 'date',
  },
  value: [Date, Array] as PropType<Date | (Date | undefined)[]>,
  activeDate: Date,
  visible: {
    type: Boolean,
    default: undefined,
  },
  isSelecting: {
    type: Boolean,
    default: undefined,
  },

  onCellClick: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  onCellMouseenter: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  'onUpdate:activeDate': [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
}

export type DatePanelProps = ExtractInnerPropTypes<typeof datePanelProps>
export type DatePanelPublicProps = ExtractPublicPropTypes<typeof datePanelProps>
export type DatePanelComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DatePanelPublicProps> & DatePanelPublicProps
>
export type DatePanelInstance = InstanceType<DefineComponent<DatePanelProps>>

export type DatePanelType = 'date' | 'week' | 'month' | 'quarter' | 'year'

// private
export const panelRowProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
}

export const panelCellProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
  cellIndex: {
    type: Number,
    required: true,
  },
}
