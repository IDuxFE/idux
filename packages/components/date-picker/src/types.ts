/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { FormSize } from '@idux/components/form'
import type { TimePickerProps } from '@idux/components/time-picker'
import type { DefineComponent, HTMLAttributes, VNode, VNodeTypes } from 'vue'

import { controlPropDef } from '@idux/cdk/forms'
import { ɵPortalTargetDef } from '@idux/cdk/portal'
import { IxPropTypes } from '@idux/cdk/utils'

const datePickerCommonProps = {
  control: controlPropDef,
  open: IxPropTypes.bool,

  allowInput: IxPropTypes.oneOfType([Boolean, IxPropTypes.oneOf(['overlay'])]),
  autofocus: IxPropTypes.bool.def(false),
  borderless: IxPropTypes.bool,
  clearable: IxPropTypes.bool,
  clearIcon: IxPropTypes.string,
  disabled: IxPropTypes.bool.def(false),
  disabledDate: IxPropTypes.func<(date: Date) => boolean>(),
  format: IxPropTypes.string,
  overlayClassName: IxPropTypes.string,
  overlayContainer: IxPropTypes.oneOfType([String, HTMLElement, IxPropTypes.func<() => string | HTMLElement>()]),
  overlayRender: IxPropTypes.func<(children: VNode[]) => VNodeTypes>(),
  readonly: IxPropTypes.bool.def(false),
  size: IxPropTypes.oneOf<FormSize>(['sm', 'md', 'lg']),
  suffix: IxPropTypes.string,
  target: ɵPortalTargetDef,
  type: IxPropTypes.oneOf<DatePickerType>(['date', 'week', 'month', 'quarter', 'year']).def('date'),

  // events
  'onUpdate:open': IxPropTypes.emit<(isOpen: boolean) => void>(),
  onClear: IxPropTypes.emit<(evt: MouseEvent) => void>(),
  onFocus: IxPropTypes.emit<(evt: FocusEvent) => void>(),
  onBlur: IxPropTypes.emit<(evt: FocusEvent) => void>(),
}

export type DatePickerCommonProps = ExtractInnerPropTypes<typeof datePickerCommonProps>
export interface DatePickerCommonBindings {
  blur: () => void
  focus: (options?: FocusOptions) => void
}

export const datePickerProps = {
  ...datePickerCommonProps,

  value: IxPropTypes.oneOfType<number | string | Date>([Number, String, Date]),
  defaultOpenValue: IxPropTypes.oneOfType<number | string | Date>([Number, String, Date]),
  placeholder: IxPropTypes.string,
  timePicker: IxPropTypes.oneOfType([Boolean, IxPropTypes.object<TimePickerProps>()]),

  'onUpdate:value': IxPropTypes.emit<(value: Date) => void>(),
  onChange: IxPropTypes.emit<(value: Date, oldValue: Date) => void>(),
  onInput: IxPropTypes.emit<(evt: Event) => void>(),
}

export type DatePickerProps = ExtractInnerPropTypes<typeof datePickerProps>
export type DatePickerPublicProps = ExtractPublicPropTypes<typeof datePickerProps>
export type DatePickerComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DatePickerPublicProps> & DatePickerPublicProps,
  DatePickerCommonBindings
>
export type DatePickerInstance = InstanceType<DefineComponent<DatePickerProps, DatePickerCommonBindings>>

export const dateRangePickerProps = {
  ...datePickerCommonProps,

  value: IxPropTypes.array<number | string | Date>(),
  defaultOpenValue: IxPropTypes.array<number | string | Date>(),
  placeholder: IxPropTypes.arrayOf(String),
  separator: IxPropTypes.oneOfType([String, IxPropTypes.vNode]),
  timePicker: IxPropTypes.oneOfType([
    Boolean,
    IxPropTypes.object<TimePickerProps>(),
    IxPropTypes.array<TimePickerProps>(),
  ]),

  'onUpdate:value': IxPropTypes.emit<(value: any[]) => void>(),
}

export type DateRangePickerProps = ExtractInnerPropTypes<typeof dateRangePickerProps>
export type DateRangePickerPublicProps = ExtractPublicPropTypes<typeof dateRangePickerProps>
export type DateRangePickerComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DateRangePickerPublicProps> & DateRangePickerPublicProps,
  DatePickerCommonBindings
>
export type DateRangePickerInstance = InstanceType<DefineComponent<DateRangePickerProps, DatePickerCommonBindings>>

export type DatePickerType = 'date' | 'week' | 'month' | 'quarter' | 'year'

// private

export const panelRowProps = {
  rowIndex: IxPropTypes.number.isRequired,
}

export const panelCellProps = {
  rowIndex: IxPropTypes.number.isRequired,
  cellIndex: IxPropTypes.number.isRequired,
  isWeek: IxPropTypes.bool,
}
