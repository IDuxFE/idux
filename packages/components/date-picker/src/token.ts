/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type { PickerControlContext } from './composables/useControl'
import type { FormatContext } from './composables/useFormat'
import type { InputEnableStatus } from './composables/useInputEnableStatus'
import type { OverlayStateContext } from './composables/useOverlayState'
import type { PickerStateContext } from './composables/usePickerState'
import type { PickerRangeControlContext } from './composables/useRangeControl'
import type { DatePickerProps, DateRangePickerProps } from './types'
import type { CommonConfig, DateConfig, DatePickerConfig } from '@idux/components/config'
import type { Locale } from '@idux/components/locales'
import type { ComputedRef, InjectionKey, Ref, Slots, VNodeTypes } from 'vue'

export interface DatePickerContext extends OverlayStateContext, FormatContext, PickerStateContext<DatePickerProps> {
  props: DatePickerProps
  slots: Slots
  common: CommonConfig
  locale: Locale
  config: DatePickerConfig
  focused: ComputedRef<boolean>
  mergedPrefixCls: ComputedRef<string>
  dateConfig: DateConfig
  inputRef: Ref<HTMLInputElement | undefined>
  inputEnableStatus: ComputedRef<InputEnableStatus>
  controlContext: PickerControlContext
  handleKeyDown: (evt: KeyboardEvent) => void
}

export interface DateRangePickerContext
  extends OverlayStateContext,
    FormatContext,
    PickerStateContext<DateRangePickerProps> {
  props: DateRangePickerProps
  slots: Slots
  common: CommonConfig
  focused: ComputedRef<boolean>
  locale: Locale
  config: DatePickerConfig
  mergedPrefixCls: ComputedRef<string>
  dateConfig: DateConfig
  inputRef: Ref<HTMLInputElement | undefined>
  inputEnableStatus: ComputedRef<InputEnableStatus>
  rangeControlContext: PickerRangeControlContext
  renderSeparator: () => VNodeTypes
  handleKeyDown: (evt: KeyboardEvent) => void
}

export const datePickerToken: InjectionKey<DatePickerContext> = Symbol('datePickerToken')
export const dateRangePickerToken: InjectionKey<DateRangePickerContext> = Symbol('dateRangePickerToken')
