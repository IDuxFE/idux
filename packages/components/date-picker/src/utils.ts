/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { DatePickerType, PresetRangeShortcut, RangeShortcutOptions } from './types'
import type { DateConfig, DateConfigType } from '@idux/components/config'

import { isFunction } from 'lodash-es'

export function convertPickerTypeToConfigType(type: DatePickerType): Exclude<DateConfigType, 'day'> {
  return type === 'datetime' ? 'date' : type
}

const oneDayDuration = 24 * 60 * 60 * 1000

export function getPresetRangeShortcutValue(
  dateConfig: DateConfig,
  presetShortcut: PresetRangeShortcut,
): () => [Date, Date] {
  const { now, startOf, endOf } = dateConfig

  switch (presetShortcut) {
    case 'today':
    default:
      return () => {
        const currentDate = now()
        return [startOf(currentDate, 'date'), endOf(currentDate, 'date')]
      }
    case 'yesterday': {
      return () => {
        const currentDate = now()
        const yesterday = new Date(currentDate.valueOf() - oneDayDuration)
        return [startOf(yesterday, 'date'), endOf(yesterday, 'date')]
      }
    }
    case 'last24h': {
      return () => {
        const currentDate = now()
        const currentHour = endOf(currentDate, 'hour')
        return [new Date(currentHour.valueOf() - oneDayDuration), currentHour]
      }
    }
    case 'last7d': {
      return () => {
        const currentDate = now()
        const endOfToday = endOf(currentDate, 'date')
        return [new Date(endOfToday.valueOf() - oneDayDuration * 7), endOfToday]
      }
    }
    case 'last30d': {
      return () => {
        const currentDate = now()
        const endOfToday = endOf(currentDate, 'date')
        return [new Date(endOfToday.valueOf() - oneDayDuration * 30), endOfToday]
      }
    }
    case 'last180d': {
      return () => {
        const currentDate = now()
        const endOfToday = endOf(currentDate, 'date')
        return [new Date(endOfToday.valueOf() - oneDayDuration * 180), endOfToday]
      }
    }
  }
}

export function extractShortcutValue(shortcut: RangeShortcutOptions): [Date, Date] | undefined {
  const value = shortcut.value

  if (!value) {
    return
  }

  return isFunction(value) ? value() : value
}
