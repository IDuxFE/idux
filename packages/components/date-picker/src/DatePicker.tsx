/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, nextTick, normalizeClass, provide, watch } from 'vue'

import { ɵOverlay } from '@idux/components/_private/overlay'
import { useDateConfig, useGlobalConfig } from '@idux/components/config'
import { useFormElement } from '@idux/components/form'

import { useControl } from './composables/useControl'
import { useFormat } from './composables/useFormat'
import { useInputEnableStatus } from './composables/useInputEnableStatus'
import { useKeyboardEvents } from './composables/useKeyboardEvents'
import { useOverlayProps } from './composables/useOverlayProps'
import { useOverlayState } from './composables/useOverlayState'
import { usePickerState } from './composables/usePickerState'
import Content from './content/Content'
import { datePickerToken } from './token'
import Trigger from './trigger/Trigger'
import { datePickerProps } from './types'

export default defineComponent({
  name: 'IxDatePicker',
  inheritAttrs: false,
  props: datePickerProps,
  setup(props, { attrs, expose, slots }) {
    const common = useGlobalConfig('common')
    const locale = useGlobalConfig('locale')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-date-picker`)
    const config = useGlobalConfig('datePicker')
    const dateConfig = useDateConfig()

    const { elementRef: inputRef, focus, blur } = useFormElement<HTMLInputElement>()

    expose({ focus, blur })

    const inputEnableStatus = useInputEnableStatus(props, config)
    const formatContext = useFormat(props, config)
    const pickerStateContext = usePickerState(props, dateConfig, formatContext.formatRef)

    const { accessor, handleChange } = pickerStateContext

    const controlContext = useControl(dateConfig, formatContext, inputEnableStatus, accessor.valueRef, handleChange)
    const { overlayOpened, setOverlayOpened } = useOverlayState(props, controlContext)
    const handleKeyDown = useKeyboardEvents(setOverlayOpened)

    const context = {
      props,
      slots,
      locale,
      config,
      mergedPrefixCls,
      dateConfig,
      inputRef,
      inputEnableStatus,
      overlayOpened,
      setOverlayOpened,
      handleKeyDown,
      controlContext,
      ...formatContext,
      ...pickerStateContext,
    }

    provide(datePickerToken, context)

    watch(overlayOpened, opened => {
      nextTick(() => {
        if (opened) {
          focus()
          inputRef.value?.dispatchEvent(new FocusEvent('focus'))
        } else {
          blur()
          inputRef.value?.dispatchEvent(new FocusEvent('blur'))
        }
      })
    })

    const renderTrigger = () => <Trigger {...attrs}></Trigger>
    const renderContent = () => <Content></Content>
    const overlayProps = useOverlayProps(context)
    const overlayClass = computed(() => normalizeClass([`${mergedPrefixCls.value}-overlay`, props.overlayClassName]))

    return () => (
      <ɵOverlay
        {...overlayProps.value}
        class={overlayClass.value}
        v-slots={{ default: renderTrigger, content: renderContent }}
        triggerId={attrs.id}
      />
    )
  },
})
