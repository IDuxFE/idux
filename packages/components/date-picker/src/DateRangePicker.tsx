/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, normalizeClass, provide, watch } from 'vue'

import { ɵOverlay } from '@idux/components/_private/overlay'
import { useDateConfig, useGlobalConfig } from '@idux/components/config'
import { useFormElement } from '@idux/components/utils'

import { useFormat } from './composables/useFormat'
import { useInputEnableStatus } from './composables/useInputEnableStatus'
import { useRangeKeyboardEvents } from './composables/useKeyboardEvents'
import { useOverlayState } from './composables/useOverlayState'
import { usePickerState } from './composables/usePickerState'
import { useRangeControl } from './composables/useRangeControl'
import RangeContent from './content/RangeContent'
import { dateRangePickerToken } from './token'
import RangeTrigger from './trigger/RangeTrigger'
import { dateRangePickerProps } from './types'

const defaultOffset: [number, number] = [0, 8]

export default defineComponent({
  name: 'IxDateRangePicker',
  inheritAttrs: false,
  props: dateRangePickerProps,
  setup(props, { attrs, expose, slots }) {
    const common = useGlobalConfig('common')
    const locale = useGlobalConfig('locale')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-date-range-picker`)
    const config = useGlobalConfig('datePicker')
    const dateConfig = useDateConfig()

    const { elementRef: inputRef, focus, blur } = useFormElement<HTMLInputElement>()

    expose({ focus, blur })

    const showFooter = computed(() => !!props.footer || !!slots.footer)

    const inputEnableStatus = useInputEnableStatus(props, config)
    const formatContext = useFormat(props, config)
    const pickerStateContext = usePickerState(props, dateConfig, formatContext.formatRef)

    const { accessor, handleChange } = pickerStateContext

    const rangeControlContext = useRangeControl(dateConfig, formatContext, inputEnableStatus, accessor.valueRef)
    const { overlayOpened, setOverlayOpened } = useOverlayState(props, rangeControlContext)
    const handleKeyDown = useRangeKeyboardEvents(rangeControlContext, showFooter, setOverlayOpened, handleChange)

    const renderSeparator = () => slots.separator?.() ?? props.separator ?? locale.dateRangePicker.separator

    provide(dateRangePickerToken, {
      props,
      slots,
      locale,
      config,
      mergedPrefixCls,
      dateConfig,
      inputRef,
      inputEnableStatus,
      overlayOpened,
      setOverlayOpened,
      rangeControlContext,
      renderSeparator,
      handleKeyDown,
      ...formatContext,
      ...pickerStateContext,
    })

    watch(overlayOpened, opened => {
      if (!opened) {
        rangeControlContext.init(true)
      }

      opened ? focus() : blur()
    })

    watch(rangeControlContext.buffer, value => {
      if (!showFooter.value) {
        handleChange(value)
      }
    })

    const target = computed(() => props.target ?? config.target ?? `${mergedPrefixCls.value}-overlay-container`)
    const renderTrigger = () => <RangeTrigger {...attrs}></RangeTrigger>
    const renderContent = () => <RangeContent></RangeContent>
    const overlayProps = { 'onUpdate:visible': setOverlayOpened }

    return () => {
      return (
        <ɵOverlay
          {...overlayProps}
          visible={overlayOpened.value}
          v-slots={{ default: renderTrigger, content: renderContent }}
          class={normalizeClass(props.overlayClassName)}
          clickOutside
          disabled={accessor.disabled.value || props.readonly}
          offset={defaultOffset}
          placement="bottomStart"
          target={target.value}
          trigger="manual"
        />
      )
    }
  },
})
