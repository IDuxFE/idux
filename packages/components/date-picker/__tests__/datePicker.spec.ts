import { MountingOptions, mount } from '@vue/test-utils'

import { renderWork } from '@tests'

import DatePicker from '../src/DatePicker'
import { DatePickerProps } from '../src/types'

describe('DatePicker', () => {
  const DatePickerMount = (options?: MountingOptions<Partial<DatePickerProps>>) =>
    mount(DatePicker, { ...(options as MountingOptions<DatePickerProps>) })

  renderWork<DatePickerProps>(DatePicker, {
    props: {},
  })

  test('v-model:value work', async () => {
    const onUpdateValue = vi.fn()
    const onChange = vi.fn()
    const wrapper = DatePickerMount({
      props: { value: new Date('2021-10-01'), 'onUpdate:value': onUpdateValue, onChange },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
