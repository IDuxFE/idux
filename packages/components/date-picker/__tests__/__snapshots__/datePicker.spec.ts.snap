// Vitest Snapshot v1

exports[`DatePicker > render work 1`] = `
"<div class=\\"ix-trigger ix-trigger-md ix-trigger-with-suffix ix-control-trigger ix-date-picker\\" tabindex=\\"-1\\" opened=\\"false\\">
  <!---->
  <div class=\\"ix-date-picker-input\\"><input class=\\"ix-date-picker-input-inner\\" autocomplete=\\"off\\" placeholder=\\"请选择日期\\" readonly=\\"\\" size=\\"12\\" value=\\"\\"></div>
  <div class=\\"ix-trigger-suffix\\"><i class=\\"ix-icon ix-icon-calendar\\" role=\\"img\\" aria-label=\\"calendar\\"></i></div>
  <!---->
</div>"
`;
